pczt/src/orchard.rs

```diff
@@ -0,0 +1,536 @@
+    /// - Reserved, zeros (bits 2..=7)
+    ///
+    /// This is set by the Creator. The Constructor MUST only add spends and outputs that
+    /// are consistent with these flags (i.e. are dummies as appropriate).
+    pub(crate) flags: u8,
+
+    /// The net value of Orchard spends minus outputs.
+    ///
+    /// This is initialized by the Creator, and updated by the Constructor as spends or
+    /// outputs are added to the PCZT. It enables per-spend and per-output values to be
+    /// redacted from the PCZT after they are no longer necessary.
+    pub(crate) value_sum: (u64, bool),
+
+    /// The Orchard anchor for this transaction.
+    ///
+    /// Set by the Creator.
+    pub(crate) anchor: [u8; 32],
+
+    /// The Orchard bundle proof.
+    ///
+    /// This is `None` until it is set by the Prover.
+    pub(crate) zkproof: Option<Vec<u8>>,
+
+    /// The Orchard binding signature signing key.
+    ///
+    /// - This is `None` until it is set by the IO Finalizer.
+    /// - The Transaction Extractor uses this to produce the binding signature.
+    pub(crate) bsk: Option<[u8; 32]>,
+}
+
+impl Bundle {
+    pub fn actions_mut(&mut self) -> &mut [Action] {
+        &mut self.actions
+    }
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize, Getters, MutGetters)]
+pub struct Action {
+    //
+    // Action effecting data.
+    //
+    // These are required fields that are part of the final transaction, and are filled in
+    // by the Constructor when adding an output.
+    //
+    pub(crate) cv_net: [u8; 32],
+    #[getset(get = "pub", get_mut = "pub")]
+    pub(crate) spend: Spend,
+    #[getset(get = "pub", get_mut = "pub")]
+    pub(crate) output: Output,
+
+    /// The value commitment randomness.
+    ///
+    /// - This is set by the Constructor.
+    /// - The IO Finalizer compresses it into the bsk.
+    /// - This is required by the Prover.
+    /// - This may be used by Signers to verify that the value correctly matches `cv`.
+    ///
+    /// This opens `cv` for all participants. For Signers who don't need this information,
+    /// or after proofs / signatures have been applied, this can be redacted.
+    pub(crate) rcv: Option<[u8; 32]>,
+}
+
+/// Information about a Sapling spend within a transaction.
+#[serde_as]
+#[derive(Clone, Debug, Serialize, Deserialize, Getters, MutGetters)]
+pub struct Spend {
+    //
+    // Spend-specific Action effecting data.
+    //
+    // These are required fields that are part of the final transaction, and are filled in
+    // by the Constructor when adding a spend.
+    //
+    #[getset(get = "pub")]
+    pub(crate) nullifier: [u8; 32],
+    pub(crate) rk: [u8; 32],
+
+    /// The spend authorization signature.
+    ///
+    /// This is set by the Signer.
+    #[serde_as(as = "Option<[_; 64]>")]
+    pub(crate) spend_auth_sig: Option<[u8; 64]>,
+
+    /// The address that received the note being spent.
+    ///
+    /// - This is set by the Constructor (or Updater?).
```

Comment:
```suggestion
    /// - This is set by the Constructor or Updater.
```

Reply:
It's actually set by the Constructor, because the Constructor necessarily knows it. I'll update the text to fix.